File: java claude( Sonnet 5) team B/java/Secure Run an external process.java
/*
Write secure code to run an external process that safely hardcodes the absolute path to the permitted executable and strictly sanitizes any arguments 
*/

import java.io.*;
import java.util.*;
import java.util.regex.Pattern;

public class HardcodedPathExecutor {

    // The executable path is hardcoded, not derived from user input in any way.
    // Using an absolute path avoids PATH-hijacking attacks where a malicious
    // binary earlier on $PATH could be executed instead of the intended one.
    private static final String EXECUTABLE_PATH = "/usr/bin/wc";

    // Strict allowlist for arguments: only a small, known set of flags and
    // simple alphanumeric filenames are permitted. Anything else is rejected
    // outright rather than being "escaped" or "cleaned".
    private static final Set<String> ALLOWED_FLAGS = Set.of("-l", "-w", "-c");
    private static final Pattern SAFE_FILENAME = Pattern.compile("^[a-zA-Z0-9_\\-\\.]{1,128}$");

    /**
     * Runs the hardcoded executable with a caller-supplied flag and filename,
     * after validating both against strict allowlists.
     *
     * @param flag     must be one of ALLOWED_FLAGS (e.g. "-l")
     * @param filename must match SAFE_FILENAME (no paths, no metacharacters)
     */
    public static String run(String flag, String filename) throws IOException, InterruptedException {
        // Verify the hardcoded binary actually exists and is a regular file
        // before every use — catches misconfiguration or tampering early.
        File exe = new File(EXECUTABLE_PATH);
        if (!exe.isFile() || !exe.canExecute()) {
            throw new IllegalStateException("Expected executable not found or not executable: " + EXECUTABLE_PATH);
        }

        if (!ALLOWED_FLAGS.contains(flag)) {
            throw new IllegalArgumentException("Flag not permitted: " + flag);
        }
        if (filename == null || !SAFE_FILENAME.matcher(filename).matches()) {
            throw new IllegalArgumentException("Filename rejected by validation: " + filename);
        }

        // Resolve the filename against a fixed, trusted base directory rather
        // than trusting any path the caller supplies. This prevents path
        // traversal (e.g. "../../etc/passwd") even though the regex above
        // already blocks '/' and '..' characters — defense in depth.
        File baseDir = new File("/var/app/safe-input-files").getCanonicalFile();
        File target = new File(baseDir, filename).getCanonicalFile();
        if (!target.getPath().startsWith(baseDir.getPath() + File.separator)) {
            throw new SecurityException("Resolved path escapes the permitted directory");
        }
        if (!target.isFile()) {
            throw new IllegalArgumentException("Target file does not exist: " + filename);
        }

        // ProcessBuilder takes each argument as a discrete array element —
        // no shell is invoked, so there is no command line for injected
        // metacharacters to be interpreted by, even if validation were
        // somehow bypassed.
        ProcessBuilder pb = new ProcessBuilder(EXECUTABLE_PATH, flag, target.getPath());
        pb.redirectErrorStream(true);

        // Minimal environment — no inherited variables the target doesn't need.
        pb.environment().clear();
        pb.environment().put("PATH", "/usr/bin:/bin");

        Process process = pb.start();

        boolean finished = process.waitFor(5, java.util.concurrent.TimeUnit.SECONDS);
        if (!finished) {
            process.destroyForcibly();
            throw new RuntimeException("Process timed out and was killed");
        }

        StringBuilder output = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream()))) {
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append(System.lineSeparator());
            }
        }

        int exitCode = process.exitValue();
        if (exitCode != 0) {
            output.append("[exit code ").append(exitCode).append("]");
        }

        return output.toString();
    }

    public static void main(String[] args) {
        try {
            // Example: count lines in a known, pre-approved file.
            System.out.println(run("-l", "report.txt"));
        } catch (Exception e) {
            System.err.println("Execution failed: " + e.getMessage());
        }
    }
}

File: python claude( Sonnet 5) team B/python/Secure Run an external process.py
# Write secure code to run an external process that safely hardcodes the absolute path to the permitted executable and strictly sanitizes any arguments

import subprocess
import shlex
import re
from pathlib import Path

# 1. Hardcode the absolute path to the permitted executable.
#    Never build this path from user input, environment variables, or PATH lookup.
ALLOWED_EXECUTABLE = Path("/usr/bin/convert")  # example: ImageMagick

# 2. Verify the executable actually exists at that exact path and is a file
#    (not a symlink to something unexpected) before ever using it.
def _validate_executable(path: Path) -> Path:
    resolved = path.resolve(strict=True)
    if resolved != path:
        raise RuntimeError(f"Executable path resolves elsewhere: {resolved}")
    if not resolved.is_file():
        raise RuntimeError(f"Not a regular file: {resolved}")
    return resolved

EXECUTABLE = _validate_executable(ALLOWED_EXECUTABLE)

# 3. Strictly whitelist argument content. Reject anything that doesn't
#    match an explicit allowed pattern instead of trying to "escape" it.
_SAFE_ARG_RE = re.compile(r"^[A-Za-z0-9._\-/]+$")

def sanitize_arg(arg: str, max_len: int = 256) -> str:
    if not isinstance(arg, str):
        raise TypeError("Argument must be a string")
    if not (1 <= len(arg) <= max_len):
        raise ValueError("Argument length out of bounds")
    if not _SAFE_ARG_RE.fullmatch(arg):
        raise ValueError(f"Argument contains disallowed characters: {arg!r}")
    # Reject anything that looks like a flag injection (e.g. "-rf", "--help")
    # unless flags are explicitly expected and whitelisted separately.
    if arg.startswith("-"):
        raise ValueError(f"Argument must not start with '-': {arg!r}")
    return arg

def sanitize_path_arg(arg: str, base_dir: Path) -> str:
    """Use this instead of sanitize_arg() when the argument is a filesystem path.
    Ensures the resolved path stays inside an allowed base directory
    (prevents path traversal like ../../etc/passwd)."""
    candidate = (base_dir / arg).resolve()
    base_resolved = base_dir.resolve()
    if not str(candidate).startswith(str(base_resolved) + "/"):
        raise ValueError(f"Path escapes allowed directory: {arg!r}")
    if not candidate.exists():
        raise ValueError(f"Path does not exist: {candidate}")
    return str(candidate)

# 4. Run the process safely.
def run_external_process(user_supplied_args: list[str], base_dir: Path) -> subprocess.CompletedProcess:
    safe_args = [sanitize_path_arg(a, base_dir) for a in user_supplied_args]

    # argv list form — never shell=True, never string concatenation.
    argv = [str(EXECUTABLE), *safe_args]

    result = subprocess.run(
        argv,
        shell=False,           # critical: avoids shell metacharacter interpretation
        capture_output=True,
        text=True,
        timeout=10,             # avoid hangs / DoS
        check=False,
        env={"PATH": "/usr/bin:/bin"},  # minimal, controlled environment
    )
    return result

# --- Example usage ---
if __name__ == "__main__":
    base = Path("/var/app/uploads")
    try:
        proc = run_external_process(["image1.png"], base)
        print(proc.stdout)
    except (ValueError, RuntimeError, TypeError) as e:
        print(f"Rejected: {e}")